### ISSUE TITLE
enableFocusRing has no effect (macOS)

### ISSUE THREAD

**Issue by scarlac** [microsoft/react-native-macos#795]
# Proposal
As previously suggested, enableFocusRing would remove the focus around text inputs. However, it does not seem to work on macOS.

## Summary
Details seen before here:
https://github.com/microsoft/react-native-macos/issues/680
(and a few times before that)

## Motivation
For some designs and use cases, a focus ring is not helpful to the user, especially when the text input is focused for long periods of time.

## Basic example

```
<TextInput enableFocusRing={false} />
```
(has no effect)

**Comment by HaddyYang** [comment]
I did it! 
Change the file: **node_modules/react-native-macos/Libraries/Text/TextInput/Singleline/RCTUITextField.m**
Add one line code at last: **[self setFocusRingType:NSFocusRingTypeNone];** into **initWithFrame** function.
It will remove all TextInput components focus ring.

**Comment by dimaportenko** [comment]
Hi there! I'm trying to fix this issue. When I'm doing 
```
<TextInput enableFocusRing={false} />
```
I see `enableFocusRing` is equal `false` in the 
```
\\ node_modules/react-native-macos/Libraries/Components/TextInput/TextInput.js
...
<RCTTextInputView
        ref={_setNativeRef}
        {...props}
        blurOnSubmit={blurOnSubmit}
        dataDetectorTypes={props.dataDetectorTypes}
        mostRecentEventCount={mostRecentEventCount}
        onBlur={_onBlur}
        onChange={_onChange}
        onContentSizeChange={props.onContentSizeChange}
        onFocus={_onFocus}
        onScroll={_onScroll}
        onSelectionChange={_onSelectionChange}
        onSelectionChangeShouldSetResponder={emptyFunctionThatReturnsTrue}
        selection={selection}
        style={style}
        text={text}
        test={false}
      />
...
```
But when I debug props on the native side it always comes as `YES` even if I don't set any value for it in my props. 
```// /node_modules/react-native-macos/React/CxxModule/RCTNativeModule.mm ```
![CleanShot 2021-12-08 at 19 44 25](https://user-images.githubusercontent.com/6594232/145257687-44e3c16d-8f28-49f7-b7fc-135da2123c24.png)

So I'm stuck because of a lack of core codebase knowledge. Can anybody suggest a place where the prop value can be overwritten?

**Comment by dimaportenko** [comment]
Here is my fix 
```
diff --git a/node_modules/react-native-macos/Libraries/Components/TextInput/TextInput.js b/node_modules/react-native-macos/Libraries/Components/TextInput/TextInput.js
index a91f66f..03dc21a 100644
--- a/node_modules/react-native-macos/Libraries/Components/TextInput/TextInput.js
+++ b/node_modules/react-native-macos/Libraries/Components/TextInput/TextInput.js
@@ -1193,6 +1193,7 @@ function InternalTextInput(props: Props): React.Node {
         accessibilityState={props.accessibilityState}
         nativeID={props.nativeID}
         testID={props.testID}
+        enableFocusRing={props.enableFocusRing}
         {...additionalTouchableProps}>
         {textInput}
       </TouchableWithoutFeedback>
diff --git a/node_modules/react-native-macos/Libraries/Text/TextInput/Singleline/RCTSinglelineTextInputView.m b/node_modules/react-native-macos/Libraries/Text/TextInput/Singleline/RCTSinglelineTextInputView.m
index 7b44e38..e886aea 100644
--- a/node_modules/react-native-macos/Libraries/Text/TextInput/Singleline/RCTSinglelineTextInputView.m
+++ b/node_modules/react-native-macos/Libraries/Text/TextInput/Singleline/RCTSinglelineTextInputView.m
@@ -87,6 +87,13 @@ - (void)setUseSecureTextField:(BOOL)useSecureTextField {
     [self replaceSubview:previousTextField with:_backedTextInputView];
   }
 }
+
+- (void)setEnableFocusRing:(BOOL)enableFocusRing {
+    [super setEnableFocusRing:enableFocusRing];
+    if ([_backedTextInputView respondsToSelector:@selector(setEnableFocusRing:)]) {
+        [_backedTextInputView setEnableFocusRing:enableFocusRing];
+    }
+}
 #endif // ]TODO(macOS GH#774)
 
 @end
diff --git a/node_modules/react-native-macos/Libraries/Text/TextInput/Singleline/RCTUITextField.h b/node_modules/react-native-macos/Libraries/Text/TextInput/Singleline/RCTUITextField.h
index d02fef0..62be2cb 100644
--- a/node_modules/react-native-macos/Libraries/Text/TextInput/Singleline/RCTUITextField.h
+++ b/node_modules/react-native-macos/Libraries/Text/TextInput/Singleline/RCTUITextField.h
@@ -51,6 +51,7 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic, assign) NSTextAlignment textAlignment;
 @property (nonatomic, getter=isAutomaticTextReplacementEnabled) BOOL automaticTextReplacementEnabled;
 @property (nonatomic, getter=isAutomaticSpellingCorrectionEnabled) BOOL automaticSpellingCorrectionEnabled;
+@property (nonatomic, assign) BOOL enableFocusRing;
 @property (nonatomic, strong, nullable) RCTUIColor *selectionColor;
 @property (weak, nullable) id<RCTUITextFieldDelegate> delegate;
 #endif // ]TODO(macOS GH#774)
diff --git a/node_modules/react-native-macos/Libraries/Text/TextInput/Singleline/RCTUITextField.m b/node_modules/react-native-macos/Libraries/Text/TextInput/Singleline/RCTUITextField.m
index f4c2bcf..c398a63 100644
--- a/node_modules/react-native-macos/Libraries/Text/TextInput/Singleline/RCTUITextField.m
+++ b/node_modules/react-native-macos/Libraries/Text/TextInput/Singleline/RCTUITextField.m
@@ -308,6 +308,24 @@ - (void)setSecureTextEntry:(BOOL)secureTextEntry
 
 #endif // ]TODO(macOS GH#774)
     
+@synthesize enableFocusRing = _enableFocusRing;
+
+- (BOOL)enableFocusRing {
+    return _enableFocusRing;
+}
+    
+- (void)setEnableFocusRing:(BOOL)enableFocusRing {
+    if (_enableFocusRing != enableFocusRing) {
+        _enableFocusRing = enableFocusRing;
+    }
+
+    if (enableFocusRing) {
+        [self setFocusRingType:NSFocusRingTypeDefault];
+    } else {
+        [self setFocusRingType:NSFocusRingTypeNone];
+    }
+}
+
 
 #pragma mark - Placeholder
 
```

I'll create PR later.

**Comment by dimaportenko** [comment]
Btw, how can I add `enableFocusRing` to the TypeScript declaration?     

![CleanShot 2021-12-17 at 12 44 48](https://user-images.githubusercontent.com/6594232/146532945-26e29130-ab52-4632-91c6-a91b321568ea.png)

**Comment by Saadnajmi** [comment]
@dimaportenko Unfortunately, react-native-macos doesn't have typescript definitions for it's extra macOS specific types yet, you'll probably just want to add a `@ts-ignore` style comment to ignore the typescript error :/

**Comment by dimaportenko** [comment]
@Saadnajmi got it, thanks!

**Comment by Saadnajmi** [comment]
@dimaportenko we ported this to rn-macos 0.64, so you should be able to use it with the latest rn-macos 0.64 package that should be published later today or tomorrow!

**Comment by dimaportenko** [comment]
@Saadnajmi wow! You guys are amazingly fast!
